lib/spack/spack/package_base.py

```diff
@@ -2794,6 +2794,17 @@ def rpath_args(self):
         """
         return " ".join("-Wl,-rpath,%s" % p for p in self.rpath)
 
+    @property
+    def non_bindable_shared_objects(self):
+        """List of shared objects that should be replaced with a different library at
+        runtime. Typically includes stub libraries like libcuda.so. When linking
+        against a library listed here, the dependent will only record its soname
+        or filename, not its absolute path, so that the dynamic linker will search
+        for it. Note: accepts both file names and directory names, for example
+        ``["libcuda.so", "stubs"]`` will ensure libcuda.so and all libraries in the
+        stubs directory are not bound by path."""
+        return []
```

Comment:
One worry for this PR is how many of these we're going to need to add -- e.g., is this needed for SYCL or for HIP as well?  What about OpenMPI?  SLURM?  Flux?

@trws are there any obvious cases that come to mind here off the top of your head?